src/Orchard.Web/Modules/Orchard.OpenId/Services/OpenIdUserManager.cs

```diff
@@ -0,0 +1,107 @@
+                throw new ArgumentNullException(nameof(user));
+            }
+
+            if (scopes == null)
+            {
+                throw new ArgumentNullException(nameof(scopes));
+            }
+
+            var identity = new ClaimsIdentity(
+                OpenIdConnectServerDefaults.AuthenticationScheme,
+                Options.ClaimsIdentity.UserNameClaimType,
+                Options.ClaimsIdentity.RoleClaimType);
+
+            // Note: the name identifier is always included in both identity and
+            // access tokens, even if an explicit destination is not specified.
+            identity.AddClaim(ClaimTypes.NameIdentifier, await GetUserIdAsync(user));
+
+            // Resolve the email address associated with the user if the underlying store supports it.
+            var email = SupportsUserEmail ? await GetEmailAsync(user) : null;
+
+            // Only add the name claim if the "profile" scope was granted.
+            if (scopes.Contains(OpenIdConnectConstants.Scopes.Profile))
+            {
+                var username = await GetUserNameAsync(user);
+
+                // Throw an exception if the username corresponds to the registered
+                // email address and if the "email" scope has not been requested.
+                if (!scopes.Contains(OpenIdConnectConstants.Scopes.Email) &&
+                    !string.IsNullOrEmpty(email) &&
+                     string.Equals(username, email, StringComparison.OrdinalIgnoreCase))
+                {
+                    throw new InvalidOperationException("The 'email' scope is required.");
+                }
+
+                identity.AddClaim(ClaimTypes.Name, username,
+                    OpenIdConnectConstants.Destinations.AccessToken,
+                    OpenIdConnectConstants.Destinations.IdentityToken);
+            }
+
+            // Only add the email address if the "email" scope was granted.
+            if (!string.IsNullOrEmpty(email) && scopes.Contains(OpenIdConnectConstants.Scopes.Email))
+            {
+                identity.AddClaim(ClaimTypes.Email, email,
+                    OpenIdConnectConstants.Destinations.AccessToken,
+                    OpenIdConnectConstants.Destinations.IdentityToken);
+            }
+
+            if (SupportsUserRole && scopes.Contains(OpenIddictConstants.Scopes.Roles))
+            {
+                foreach (var role in await GetRolesAsync(user))
+                {
+                    identity.AddClaim(identity.RoleClaimType, role,
+                        OpenIdConnectConstants.Destinations.AccessToken,
+                        OpenIdConnectConstants.Destinations.IdentityToken);
+                    foreach (var claim in await _roleManager.GetClaimsAsync(await _roleManager.FindByNameAsync(role)))
+                    {
+                        identity.AddClaim(claim.Type, claim.Value, OpenIdConnectConstants.Destinations.IdentityToken);
+                    }                   
+                }
```

Comment:
But don't is a bad practice to put those claims in AccessToken because it is going to be passed on each request?
I would prefer that middleware that process the access token adds those claims based on the userid of the token referenced by access token. How can I do that?